tests/test-class-wp-offline-page-ui.php

```diff
@@ -0,0 +1,326 @@
+		$trashed_post_id = $this->factory()->post->create( array( 'post_status' => 'trash' ) );
+		$this->assertEquals( null, $this->instance->sanitize_callback( $trashed_post_id ) );
+		$this->assertEquals(
+			array_merge(
+				$settings_error,
+				array(
+					'message' => 'The currently offline page is in the trash. Please select or create one or <a href="edit.php?post_status=trash&post_type=page">restore the current page</a>.',
+				)
+			),
+			reset( $wp_settings_errors )
+		);
+
+		$wp_settings_errors = array(); // WPCS: global override OK.
+
+		// The argument passed to the sanitize_callback() is a valid page ID, so it should return it.
+		$valid_post_id = $this->factory()->post->create( array( 'post_type' => 'page' ) );
+		$this->assertEquals( $valid_post_id, $this->instance->sanitize_callback( $valid_post_id ) );
+		$this->assertEquals( array(), $wp_settings_errors );
+	}
+
+	/**
+	 * Test add_settings_field.
+	 *
+	 * @covers WP_Offline_Page::add_settings_field()
+	 */
+	public function test_add_settings_field() {
+		global $wp_settings_fields;
+
+		$this->instance->add_settings_field();
+		$this->assertEquals(
+			array(
+				'id'       => WP_Offline_Page_UI::SETTING_ID,
+				'title'    => 'Page displays when offline',
+				'callback' => array( $this->instance, 'render_settings' ),
+				'args'     => array(),
+			),
+			$wp_settings_fields[ WP_Offline_Page_UI::OPTION_GROUP ]['default'][ WP_Offline_Page_UI::SETTING_ID ]
+		);
+	}
+
+	/**
+	 * Test render_settings.
+	 *
+	 * @covers WP_Offline_Page::render_settings()
+	 */
+	public function test_render_settings() {
+		// Check with there are no pages.
+		ob_start();
+		$this->instance->render_settings();
+		$output = ob_get_clean();
+		$this->assertNotContains( 'Select an existing page:', $output );
+		$this->assertContains( 'Create a new offline page', $output );
+		$this->assertContains( 'This page is for the Progressive Web App (PWA)', $output );
+
+		// Check when there are pages.
+		$number_pages = 10;
+		$page_ids     = array();
+		for ( $i = 0; $i < $number_pages; $i ++ ) {
+			$page_ids[] = $this->factory()->post->create( array( 'post_type' => 'page' ) );
+		}
+		ob_start();
+		$this->instance->render_settings();
+		$output = ob_get_clean();
+		$this->assertContains( 'Select an existing page:', $output );
+
+		// All of the pages should appear in the <select> element from wp_dropdown_pages().
+		foreach ( $page_ids as $page_id ) {
+			$this->assertContains( '<option class="level-0" value="' . $page_id . '">', $output );
+		}
+
+		$this->assertContains(
+			'<a href="' . admin_url( 'options-reading.php?action=create-offline-page' ) . '">create a new one</a>',
+			$output
+		);
+
+		// Check that it excludes the configured static pages.
+		update_option( 'page_on_front', (int) $page_ids[0] );
+		update_option( 'page_for_posts', (int) $page_ids[1] );
+		update_option( 'wp_page_for_privacy_policy', (int) $page_ids[2] );
```

Comment:
This string should instead be `\WP_Offline_Page::OPTION_NAME`.

Reply:
Privacy policy page is a special page too.

Reply:
@hellofromtonya Oh, my mistake! I mistakenly read `wp_page_for_privacy_policy` as `wp_page_for_offline_page` or something. I actually hadn't seen the `wp_page_for_privacy_policy` option before, so I mistakenly glossed over it.